### ISSUE TITLE
Enable composition of typed actors via inheritance

### ISSUE THREAD

**Issue by Neverlord** [actor-framework/actor-framework#425]
Currently, there is no way at a class-hierarchy level to combine actor behaviors. Typed actors have the potential to allow such composition. Here is a motivating example:

```cpp
using i3_actor = typed_actor<replies_to<int, int, int>::with<int>>;

using d_actor = typed_actor<replies_to<double>::with<double, double>>;

using foo_actor = i3_actor::extend_with<d_actor>;

// implement combined foo actor directly
class foo_actor_impl : public foo_actor::base {
public:
  int operator()(const int& x, const int& y, const int& z) override {
    return x + y + z;
  }

  std::tuple<double, double> operator()(const double& x) override {
    return std::make_tuple(x, x);
  }
};

class i3_actor_impl : public i3_actor::base {
public:
  int operator()(const int& x, const int& y, const int& z) override {
    return x + y + z;
  }
};

class d_actor_impl : public d_actor::base {
public:
  std::tuple<double, double> operator()(const double& x) override {
    return std::make_tuple(x, x);
  }
};

// implement combined foo actor by reusing i3_actor_impl and d_actor_impl
class foo_actor_impl2 : public i3_actor_impl, public d_actor_impl {
  // ... could override call operators ...
};
```

Pros:
- Developers can re-use actor behaviors *easily*
- Developers can override individual message handlers of parents

Cons:
- Typed actors are implemented as abstract classes by user
- CAF needs to inherit from user-provided types
- Base classes provide all `operator()`s with const-only arguments

The `operator()`s could also always return `maybe<T>` instead of `T`, but that's detail.

The first step to achieve the proposed syntax is to reshape CAF's class hierarchy:
- The class `local_actor` needs to become a pure interface
- The current `local_actor` class would then be renamed to `local_actor_base`
- Actors would always use `local_actor` as virtual base

This allows us to use `send`, `request`, `spawn`, etc. in each typed actor base without having a "final" actor type. The second step is to improve `spawn` to recognize "incomplete" actor types and to automatically inherit from this type for mixing in actual data members.

**Comment by Neverlord** [comment]
Another idea to this topic: compose *states* instead of actual actors. This is equally powerful but comes with much less impact on the API of CAF.

**Comment by Neverlord** [comment]
I played a bit around with this feature and have a working draft in my local branch. Here's a snippet how composable states might look like:

```cpp
using sink = typed_actor<reacts_to<put_atom, int>>;

using source = typed_actor<replies_to<get_atom>::with<int>>;

class sink_state : public composable_state<sink> {
public:
  result<void> operator()(int x) {
    x_ = x;
    return none;
  }
protected:
  int x_;
};

class io_state : public composed_state<sink_state, composable_state<source>> {
public:
  result<int> operator()(get_atom) {
    return x_;
  }
};

int main() {
  actor_system sys;
  // io is of type sink::extend_with<source>
  auto io = sys.spawn<io_state>();
  // ...
}
```

The `result<T>` type represents all possible return values from message handlers:
- an actual value of `T`
- errors, either converted from an error enum or via `make_error`
- a `delegated<T>` returned from `delegate(...)`
- a continue helper returned from `request(...).then(...)`

When combining states with overlapping message handlers, `composed_state` will set all ambiguous handlers to pure virtual. This forces users to resolve "conflicts" on their own as opposed to picking one of the handlers over the other (which most likely will result in unexpected behavior).

**Comment by Neverlord** [comment]
Shipped in experimental state in 0.15
